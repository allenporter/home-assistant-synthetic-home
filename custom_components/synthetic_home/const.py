"""Constants for Synthetic Home."""

DOMAIN = "synthetic_home"
DEFAULT_NAME = "Synthetic Home"
CONF_FILENAME = "config_filename"

# Used for loading initial state of an entity and it persists across config
# entry reloads. Updated with service calls.
DATA_DEVICE_STATES = "device_states"

ATTR_AREA_NAME = "area"
ATTR_DEVICE_NAME = "device"
ATTR_DEVICE_STATE_KEY = "device_state_key"
ATTR_CONFIG_ENTRY_ID = "config_entry_id"
