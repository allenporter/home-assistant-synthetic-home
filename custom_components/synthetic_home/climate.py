"""Climate platform for Synthetic Home."""

from typing import Any
import logging


from homeassistant.config_entries import ConfigEntry
from homeassistant.components.climate import (
    ATTR_HVAC_MODE,
    ATTR_TARGET_TEMP_HIGH,
    ATTR_TARGET_TEMP_LOW,
    ClimateEntity,
    ClimateEntityFeature,
    HVACAction,
    HVACMode,
    DOMAIN as CLIMATE_DOMAIN,
)
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.core import HomeAssistant
from homeassistant.const import ATTR_TEMPERATURE

from .const import DOMAIN
from .entity import SyntheticEntity
from .model import ParsedEntity

_LOGGER = logging.getLogger(__name__)

FAN_MODES = ["low", "high", "off"]


async def async_setup_entry(
    hass: HomeAssistant, entry: ConfigEntry, async_add_devices: AddEntitiesCallback
) -> None:
    """Set up climate platform."""

    synthetic_home = hass.data[DOMAIN][entry.entry_id]

    _LOGGER.debug("hvac_mode=%s", synthetic_home.entities)

    async_add_devices(
        SyntheticHomeClimate(entity, **entity.attributes)
        for entity in synthetic_home.entities
        if entity.platform == CLIMATE_DOMAIN
    )


class SyntheticHomeClimate(SyntheticEntity, ClimateEntity):
    """Representation of a demo climate device."""

    _attr_should_poll = False
    _attr_fan_modes = FAN_MODES
    _attr_fan_mode = None
    _attr_current_fan_mode = "off"

    def __init__(
        self,
        entity: ParsedEntity,
        hvac_modes: list[HVACMode],
        *,
        supported_features: ClimateEntityFeature | None = None,
        unit_of_measurement: str | None = None,
        current_temperature: float | None = None,
        target_temperature: float | None = None,
        hvac_mode: HVACMode | None = None,
        hvac_action: HVACAction | None = None,
    ) -> None:
        """Initialize the climate device."""
        super().__init__(entity)
        if supported_features is not None:
            self._attr_supported_features = ClimateEntityFeature(0) | supported_features
        self._attr_target_temperature = target_temperature
        self._attr_target_temperature_high = None
        self._attr_target_temperature_low = None
        self._attr_current_temperature = current_temperature
        self._attr_hvac_action = hvac_action
        self._attr_hvac_mode = hvac_mode
        if hvac_modes is not None:
            self._attr_hvac_modes = hvac_modes
        if unit_of_measurement is not None:
            self._attr_temperature_unit = unit_of_measurement

    async def async_set_temperature(self, **kwargs: Any) -> None:
        """Set new target temperatures."""
        if kwargs.get(ATTR_TEMPERATURE) is not None:
            self._attr_target_temperature = kwargs.get(ATTR_TEMPERATURE)
        if (
            kwargs.get(ATTR_TARGET_TEMP_HIGH) is not None
            and kwargs.get(ATTR_TARGET_TEMP_LOW) is not None
        ):
            self._attr_target_temperature_high = kwargs.get(ATTR_TARGET_TEMP_HIGH)
            self._attr_target_temperature_low = kwargs.get(ATTR_TARGET_TEMP_LOW)
        if (hvac_mode := kwargs.get(ATTR_HVAC_MODE)) is not None:
            self._attr_hvac_mode = hvac_mode
        self.async_write_ha_state()

    async def async_set_fan_mode(self, fan_mode: str) -> None:
        """Set new fan mode."""
        self._attr_current_fan_mode = fan_mode
        self.async_write_ha_state()

    async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
        """Set new operation mode."""
        self._attr_hvac_mode = hvac_mode
        self.async_write_ha_state()

    async def async_set_preset_mode(self, preset_mode: str) -> None:
        """Update preset_mode on."""
        self._attr_preset = preset_mode
        self.async_write_ha_state()
