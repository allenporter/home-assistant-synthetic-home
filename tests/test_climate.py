"""Test Synthetic Home climate entity."""

import pytest

from homeassistant.const import Platform
from homeassistant.components.climate import (
    SERVICE_SET_HVAC_MODE,
    DOMAIN as CLIMATE_DOMAIN,
    ATTR_HVAC_MODE,
)
from homeassistant.const import ATTR_ENTITY_ID
from homeassistant.core import HomeAssistant
from homeassistant.helpers import device_registry as dr

from .conftest import FIXTURES

TEST_ENTITY = "climate.family_room"


@pytest.fixture(name="platforms")
def mock_platforms() -> list[Platform]:
    """Set up switch platform."""
    return [Platform.CLIMATE]


@pytest.mark.parametrize(
    ("config_yaml_fixture"),
    [(f"{FIXTURES}/hvac-example.yaml")],
)
async def test_climate_hvac_entity(
    hass: HomeAssistant, setup_integration: None
) -> None:
    """Test an HVAC device."""

    state = hass.states.get(TEST_ENTITY)
    assert state
    assert state.state == "cool"
    assert state.attributes == {
        "friendly_name": "Family Room",
        "current_temperature": 16.7,
        "fan_mode": None,
        "fan_modes": ["low", "high", "off"],
        "hvac_action": "cooling",
        "hvac_modes": ["off", "cool", "heat", "auto"],
        "max_temp": 35,
        "min_temp": 7,
        "supported_features": 394,
        "target_temp_high": None,
        "target_temp_low": None,
    }
    attributes = dict(state.attributes)

    await hass.services.async_call(
        CLIMATE_DOMAIN,
        SERVICE_SET_HVAC_MODE,
        {ATTR_ENTITY_ID: TEST_ENTITY, ATTR_HVAC_MODE: "heat"},
        blocking=True,
    )
    await hass.async_block_till_done()
    state = hass.states.get(TEST_ENTITY)
    assert state
    assert state.state == "heat"
    assert state.attributes == attributes

    device_registry = dr.async_get(hass)
    assert len(device_registry.devices) == 1
    device = next(iter(device_registry.devices.values()))
    assert device.suggested_area == "Family room"
    assert device.name == "Family Room"
    assert device.manufacturer == "Nest"
    assert device.sw_version == "1.0.0"


@pytest.mark.parametrize(
    ("config_yaml_fixture", "test_entity"),
    [(f"{FIXTURES}/heat-pump-example.yaml", "climate.family_room")],
)
async def test_heat_pump(
    hass: HomeAssistant,
    setup_integration: None,
    test_entity: str,
) -> None:
    """Test a heat pump climate entity."""

    state = hass.states.get(test_entity)
    assert state
    assert state.state == "heat"
    assert state.attributes == {
        "friendly_name": "Family Room",
        "current_temperature": 15.6,
        "fan_mode": None,
        "fan_modes": ["low", "high", "off"],
        "hvac_action": "heating",
        "hvac_modes": ["heat", "off"],
        "max_temp": 35,
        "min_temp": 7,
        "supported_features": 392,
    }
    attributes = dict(state.attributes)

    await hass.services.async_call(
        CLIMATE_DOMAIN,
        SERVICE_SET_HVAC_MODE,
        {ATTR_ENTITY_ID: TEST_ENTITY, ATTR_HVAC_MODE: "heat"},
        blocking=True,
    )
    await hass.async_block_till_done()
    state = hass.states.get(TEST_ENTITY)
    assert state
    assert state.state == "heat"
    assert state.attributes == attributes
