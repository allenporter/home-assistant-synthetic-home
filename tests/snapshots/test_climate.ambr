# serializer version: 1
# name: test_hvac_restorable_attributes[tests/fixtures/hvac-example.yaml-cooling]
  tuple(
    'cool',
    ReadOnlyDict({
      'current_temperature': 25,
      'fan_mode': None,
      'fan_modes': list([
        'low',
        'high',
        'off',
      ]),
      'friendly_name': 'Family Room',
      'hvac_action': 'cooling',
      'hvac_modes': list([
        <HVACMode.OFF: 'off'>,
        <HVACMode.COOL: 'cool'>,
        <HVACMode.HEAT: 'heat'>,
        <HVACMode.AUTO: 'auto'>,
      ]),
      'max_temp': 35,
      'min_temp': 7,
      'supported_features': <ClimateEntityFeature: 394>,
      'target_temp_high': None,
      'target_temp_low': None,
    }),
  )
# ---
# name: test_hvac_restorable_attributes[tests/fixtures/hvac-example.yaml-off]
  tuple(
    'cool',
    ReadOnlyDict({
      'current_temperature': 22,
      'fan_mode': None,
      'fan_modes': list([
        'low',
        'high',
        'off',
      ]),
      'friendly_name': 'Family Room',
      'hvac_action': 'off',
      'hvac_modes': list([
        <HVACMode.OFF: 'off'>,
        <HVACMode.COOL: 'cool'>,
        <HVACMode.HEAT: 'heat'>,
        <HVACMode.AUTO: 'auto'>,
      ]),
      'max_temp': 35,
      'min_temp': 7,
      'supported_features': <ClimateEntityFeature: 394>,
      'target_temp_high': None,
      'target_temp_low': None,
    }),
  )
# ---
# name: test_hvac_restorable_attributes[tests/fixtures/hvac-example.yaml-very-low]
  tuple(
    'cool',
    ReadOnlyDict({
      'current_temperature': 25,
      'fan_mode': None,
      'fan_modes': list([
        'low',
        'high',
        'off',
      ]),
      'friendly_name': 'Family Room',
      'hvac_action': 'cooling',
      'hvac_modes': list([
        <HVACMode.OFF: 'off'>,
        <HVACMode.COOL: 'cool'>,
        <HVACMode.HEAT: 'heat'>,
        <HVACMode.AUTO: 'auto'>,
      ]),
      'max_temp': 35,
      'min_temp': 7,
      'supported_features': <ClimateEntityFeature: 394>,
      'target_temp_high': None,
      'target_temp_low': None,
    }),
  )
# ---
